### PR TITLE
gh-109587: Allow "precompiled" perf-trampolines to largely mitigate the cost of enabling perf-trampolines

### PR DESCRIPTION
**Pull request by gsallam**
This pull request implements a proposal in #109587 to allow precompiled perf-trampolines. This would mitigate the costs of enabling perf-trampolines, such as the disk IO and memory overhead.

The proposal introduces two new C-API functions:

- `PyUnstable_PerfTrampoline_CompileCode()`: Creates a new trampoline and registers it.
- `PyUnstable_PerfTrampoline_SetPersistAfterFork()`: Controls whether to re-initialize trampolines in child processes.

These functions can be used by extension modules to initialize trampolines eagerly, after the application is "warmed up". This would make it possible to have perf-trampoline running in an always-enabled fashion.

Benefits for a forked multiprocess model:
- Reduces the CPU overhead of enabling perf-trampolines.
- Makes it possible to have perf-trampoline running in an always-enabled fashion.

* Issue: gh-109587

### CODE REVIEW

File: Misc/stable_abi.toml [hunk 1]
```diff
@@ -2460,3 +2460,7 @@
     added = '3.13'
 [function.PyMapping_HasKeyStringWithError]
     added = '3.13'
+[function.PyUnstable_PerfTrampoline_CompileCode]
+    added = '3.13'
+[function.PyUnstable_PerfTrampoline_SetPersistAfterFork]
+    added = '3.13'
```
Comment (brandtbucher) [inline Misc/stable_abi.toml]: Just a drive-by comment... I'm not sure why we would want to include unstable functions like these in the stable ABI. Is there a particular reason why they're being added?

Comment (gsallam) [reply Misc/stable_abi.toml]: Sorry for the confusion. I accidentally included these functions in the stable ABI while trying to fix a failing test. I have since removed them.

File: Lib/test/test_stable_abi_ctypes.py [hunk 2]
```diff
@@ -805,6 +805,8 @@ def test_windows_feature_macros(self):
     "PyUnicode_Translate",
     "PyUnicode_Type",
     "PyUnicode_WriteChar",
+    "PyUnstable_PerfTrampoline_CompileCode",
+    "PyUnstable_PerfTrampoline_SetPersistAfterFork",
```
Comment (brandtbucher) [inline Lib/test/test_stable_abi_ctypes.py]: Here too?

File: Include/cpython/ceval.h [hunk 3]
```diff
@@ -23,3 +23,7 @@ _PyEval_RequestCodeExtraIndex(freefunc f) {
 
 PyAPI_FUNC(int) _PyEval_SliceIndex(PyObject *, Py_ssize_t *);
 PyAPI_FUNC(int) _PyEval_SliceIndexNotNone(PyObject *, Py_ssize_t *);
+#ifndef Py_LIMITED_API
```
Comment (carljm) [inline Include/cpython/ceval.h]: This `#ifndef` is not necessary; this entire header is already only included `(from `Include/ceval.h` if `Py_LIMITED_API` is not defined.

**Comment by czardoz** [comment]
@pablogsal , @gpshead  Would love to hear your thoughts on this (whenever you get a chance to take a look) :)

**Comment by czardoz** [comment]
Ok I resolved the merge conflict, but I think the new MacOS failure is unrelated.

File: Python/sysmodule.c [hunk 4]
```diff
@@ -2374,6 +2374,44 @@ PyAPI_FUNC(void) PyUnstable_PerfMapState_Fini(void) {
...
+    char buf[4096];
+    PyThread_acquire_lock(perf_map_state.map_lock, 1);
+    while (1) {
+        size_t bytes_read = fread(buf, 1, sizeof(buf), from);
+        size_t bytes_written = fwrite(buf, 1, bytes_read, perf_map_state.perf_map);
+        fflush(perf_map_state.perf_map);
+
+        if (bytes_read < sizeof(buf) && feof(from)) {
+            fclose(from);
+            PyThread_release_lock(perf_map_state.map_lock);
```
Comment (pablogsal) [inline Python/sysmodule.c]: You can refactor this two outside of the `while`, and make a `goto exit` label that jumps to there. Thay way is easier to see that the lock is always released. You will need to assign a result variable to 0/-1 accordingly

File: Python/sysmodule.c [hunk 5]
```diff
@@ -2374,6 +2374,44 @@ PyAPI_FUNC(void) PyUnstable_PerfMapState_Fini(void) {
...
+        if(ret != 0){
+            return ret;
+        }
+    }
+    char buf[4096];
+    PyThread_acquire_lock(perf_map_state.map_lock, 1);
+    while (1) {
+        size_t bytes_read = fread(buf, 1, sizeof(buf), from);
+        size_t bytes_written = fwrite(buf, 1, bytes_read, perf_map_state.perf_map);
+        fflush(perf_map_state.perf_map);
```
Comment (pablogsal) [inline Python/sysmodule.c]: Technically `fflush` can fail, but not sure how we want to handle that (probably we can ignore it).

File: Python/sysmodule.c [hunk 6]
```diff
@@ -2374,6 +2374,44 @@ PyAPI_FUNC(void) PyUnstable_PerfMapState_Fini(void) {
...
 
+PyAPI_FUNC(int) PyUnstable_CopyPerfMapFile(const char* parent_filename) {
+#ifndef MS_WINDOWS
+    FILE* from = fopen(parent_filename, "r");
+    if (!from) {
+        return -1;
+    }
+    if (perf_map_state.perf_map == NULL) {
+        int ret = PyUnstable_PerfMapState_Init();
+        if(ret != 0){
```
Comment (pablogsal) [inline Python/sysmodule.c]:
```suggestion
        if (ret != 0){
```

File: Python/perf_trampoline.c [hunk 7]
```diff
@@ -448,16 +466,34 @@ _PyPerfTrampoline_Fini(void)
...
-    int was_active = _PyIsPerfTrampolineActive();
-    _PyPerfTrampoline_Fini();
     PyUnstable_PerfMapState_Fini();
-    if (was_active) {
-        _PyPerfTrampoline_Init(1);
+    if (persist_after_fork){
+        char filename[256];
+        pid_t parent_pid = getppid();
+        snprintf(filename, sizeof(filename), "/tmp/perf-%d.map", parent_pid);
+        if(PyUnstable_CopyPerfMapFile(filename) != 0){
```
Comment (pablogsal) [inline Python/perf_trampoline.c]:
```suggestion
        if (PyUnstable_CopyPerfMapFile(filename) != 0){
```

**Review by pablogsal — CHANGES_REQUESTED** [review]
Some minor nits

**Comment by gsallam** [comment]
I have made the requested changes; please review again

**Review by pablogsal — COMMENTED** [review]
Thanks a lot! Could you add a small test to the perf_profiler test file that checks that this works? Give it a go and if you find it complicated to test, mention it here and we can think how to proceed from there

**Comment by pablogsal** [comment]
This file: https://github.com/python/cpython/blob/main/Lib/test/test_perf_profiler.py

**Comment by gsallam** [comment]
@pablogsal I have added unit tests. Please take a look when you get a chance.

**Review by pablogsal — APPROVED** [review]

**Review by pablogsal — APPROVED** [review]
LGTM

Seems that we have some conflicts that we need to address before landing, @gsallam can you take a look?

**Comment by pablogsal** [comment]
@gsallam We still have some failures in the Windows and macOS CI. I think this is because the symbol doesn't exist in those cases:

```
_testinternalcapi.obj : error LNK2019: unresolved external symbol __imp_PyUnstable_PerfTrampoline_CompileCode referenced in function compile_perf_trampoline_entry [D:\a\cpython\cpython\PCbuild\_testinternalcapi.vcxproj]
D:\a\cpython\cpython\PCbuild\amd64\_testinternalcapi_d.pyd : fatal error LNK1120: 1 unresolved externals [D:\a\cpython\cpython\PCbuild\_testinternalcapi.vcxproj]
    0 Warning(s)
    2 Error(s)
```

**Comment by pablogsal** [comment]
We still have some failing tests it seems

**Comment by czardoz** [comment]
This approach is likely quite silly, but I don't have a Windows/Mac machine handy, so I'm kicking off these builds to test/debug.

**Comment by czardoz** [comment]
I will look at this over the weekend on my Windows box, this trial-and-error is not an optimal way :)

**Comment by pablogsal** [comment]
> I will look at this over the weekend on my Windows box, this trial-and-error is not an optimal way :)

Don't worry, I pushed a commit that should fix it

**Comment by czardoz** [comment]
Thanks a lot @pablogsal! :smile:
